### ISSUE TITLE
What about sparse matrices?

### ISSUE THREAD

**Issue by lukashes** [gonum/matrix#301]
I have big-size matrices, but it too sparsed. I do not have enough memory, and I do not see other libraries to do it.

**Comment by btracey** [comment]
We understand the importance of sparse matrices, but there is a significant coding burden in creating them. For one, there is a large diversity in sparse kinds. We are close to stabilizing mat64 and so are maybe in a position tothink about adding them, but there are no explicit plans as far as I know.

**Comment by kortschak** [comment]
I have spent quite some time thinking about a plan to introduce sparse matrix handling - it is not trivial, both in the implementation of the matrix types and also in the threading it into the mat64 operation model we use. I do want to do it though.

What operations are you particularly interested in?

**Comment by lukashes** [comment]
I need basic operations:
1. Multiplying matrices or elements of matrices
2. Growing matrices
3. Normalizing vectors

**Comment by kortschak** [comment]
1. sparse x sparse or dense x sparse
2. growing in one or two dimensions. if one, which?

**Comment by lukashes** [comment]
I am going to use only sparse matrices (sparse x sparse) and growing by two dimensions.

**Comment by kortschak** [comment]
Mmm. That's not likely to be handled for a long time. Sparse x sparse is both difficult and unlikely to be pay back well. The normal approach to sparse operations is to use a scatter which is easy to to apply effieciently to a dense, but not to a sparse. In addition to that sparsity tends to reduce, so sparse x sparse is likely to result in a dense anyway.

What is the domain you are working in, what number of non-zero elements do you expect (and what arrangement) and how fast do you need this to be? In the first instance I would suggest that you implement a sparse type on top of a map[struct{int, int}]float64. This will be slow, but it will work.

**Comment by vladimir-ch** [comment]
I will interfere a bit. I would also like to have sparse matrices in gonum, I think that eventually we will have to have them but I also agree that it is not trivial. I started to implement my own tiny sparse package but due to the lack of free time I did not get far.

To give a sample of different (much more basic) sparse operations: at first I would be happy just being able to build a fixed sparse matrix and compute (transposed) sparse matrix times dense vector product. Then I could write simple CG and GMRES solvers with which I could then solve my toy PDE models. As a next step, I would like to be able to modify values in the sparse matrix without changing its sparsity structure. That would keep me happy for a long time.

**Comment by btracey** [comment]
The trick though is that there are so many kinds of sparse matrix. At least in my world of PDEs, it's almost always block-structured rather than random access. GMRES is then coded to operate knowing the block structure.

**Comment by kortschak** [comment]
>To give a sample of different (much more basic) sparse operations: at first I would be happy with just being able to build a fixed sparse matrix and compute (transposed) sparse matrix times dense vector product.

This is trivial. You can see how this works in the sparse PageRank function in graph/network.

>Then I could write simple CG and GMRES solvers with which I could then solve my toy PDE models. As a next step, I would like to be able to modify values in the sparse matrix without changing its sparsity structure.

This is also relatively straightforward. Depending on your use model either a iteration over a row/column (depending on your storage) to find the correct index, or binary search on a sorted rox/column.

For reference, there is [CSparse](http://people.sc.fsu.edu/~jburkardt/c_src/csparse/csparse.html) and [Sparse BLAS](http://math.nist.gov/spblas/) as competitors for models to use. Sparse BLAS is really where I would like to go. Neither is trivial.

**Comment by vladimir-ch** [comment]
Yes, both are straightforward and I have them together with a CG solver. I was trying to think of a nice interface that fits with mat64 but it was not so clear to me. That's what I mean by non-trivial. I read the Sparse BLAS document and reference code in detail and tried to use it for inspiration but its usage of matrix handles as integers would have to replaced with something more Go-like.

**Comment by kortschak** [comment]
A pointer to a sparse BLAS type (types presumably - satisfying a sparse BLAS interface - rather than our blas64 method-less types) would be appropriate. The GC handles the end sparse BLAS function and a NewT handles each of the begin. I think that is all fairly simple (though tedious in the implementation).

The complexity arises in the combinatoric interaction between types both in the sparse BLAS and in mat64.

**Comment by dutradda** [comment]
any progress here, guys?

**Comment by rlouf** [comment]
I’m a beginner gopher, but I would like to avoid adding a layer of python to he existing Go backend just to have sparse matrices. If this is still something you would like to see implemented, and are willing to give me a few pointers I might give it a go (!).

**Comment by kortschak** [comment]
This repository if deprecated and frozen, so I'll move this to https://github.com/gonum/gonum/issues/367, but the overall plan would to:

1. choose a sparse matrix implementation model - I think [Sparse BLAS](http://math.nist.gov/spblas/) is probably the right choice, so reading up its documentation would be a good place to start (The [CSparse](http://people.sc.fsu.edu/%7Ejburkardt/c_src/csparse/csparse.html) book would also be worth reading as it is intended as a teaching model).
2. implement the sparse model
3. design the interaction between the sparse model and mat types - this is reasonably constrained, but should involve a fair bit of discussion with/between us.
4. implement the interactions.
